### ISSUE TITLE
Type inferring bug

### ISSUE THREAD

**Issue by auderson** [numba/numba#6903]
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

It seems I found a type inferring issue. Here is an example to reproduce it:

```
@njit
def inner(x, x_=None):
    if x_ is not None:
        y = x_
    else:
        y = x + 1
    return np.add(y, 1)

@njit
def test(a):
    if a > 0:
        b = np.arange(5.)
    else:
        b = None
    return inner(np.arange(5.), b)

>>> test(1)
```
this produces the following error:
```
Traceback (most recent call last):
  File "/home/auderson/miniconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3437, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-203-81ccd7593ad9>", line 1, in <module>
    test(1)
  File "/home/auderson/miniconda3/lib/python3.8/site-packages/numba/core/dispatcher.py", line 420, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/auderson/miniconda3/lib/python3.8/site-packages/numba/core/dispatcher.py", line 361, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<ufunc 'add'>) found for signature:
 
 >>> add(OptionalType(array(float64, 1d, C)) i.e. the type 'array(float64, 1d, C) or None', Literal[int](1))
 
There are 2 candidate implementations:
      - Of which 1 did not match due to:
      Overload in function 'Numpy_rules_ufunc.generic': File: numba/core/typing/npydecl.py: Line 96.
        With argument(s): '(OptionalType(array(float64, 1d, C)) i.e. the type 'array(float64, 1d, C) or None', int64)':
       Rejected as the implementation raised a specific error:
         TypingError: can't resolve ufunc add for types (OptionalType(array(float64, 1d, C)), int64)
  raised from /home/auderson/miniconda3/lib/python3.8/site-packages/numba/core/typing/npydecl.py:102
      - Of which 1 did not match due to:
      Overload in function 'Numpy_rules_ufunc.generic': File: numba/core/typing/npydecl.py: Line 96.
        With argument(s): '(OptionalType(array(float64, 1d, C)) i.e. the type 'array(float64, 1d, C) or None', Literal[int](1))':
       Rejected as the implementation raised a specific error:
         TypingError: can't resolve ufunc add for types (OptionalType(array(float64, 1d, C)), Literal[int](1))
  raised from /home/auderson/miniconda3/lib/python3.8/site-packages/numba/core/typing/npydecl.py:102
During: resolving callee type: Function(<ufunc 'add'>)
During: typing of call at <ipython-input-202-1466296e42c7> (7)
File "<ipython-input-202-1466296e42c7>", line 7:
def inner(x, x_=None):
    <source elided>
        y = x + 1
    return np.add(y, 1)
    ^
During: resolving callee type: type(CPUDispatcher(<function inner at 0x7f29bc1adaf0>))
During: typing of call at <ipython-input-202-1466296e42c7> (14)
During: resolving callee type: type(CPUDispatcher(<function inner at 0x7f29bc1adaf0>))
During: typing of call at <ipython-input-202-1466296e42c7> (14)
File "<ipython-input-202-1466296e42c7>", line 14:
def test(a):
    <source elided>
        b = None
    return inner(np.arange(5.), b)
    ^
```

as you can see, variable `b` from `test` is `OptionalType(array(float64, 1d, C))`, which caused the error. 
but in `inner`, there is a check for `None`, so `y` ‘s type should not be `OptionalType(array(float64, 1d, C))` any more.

if change `np.add` to `+`, it will cause a `LoweringError`

if it's not easy to change how type inferring works, do I have other ways to deal with `OptionalType(array(float64, 1d, C))` ? i.e. how can I make it a finite array type  inside the `inner` function (I do need the argument to be **optional**)?

my numba version is '0.53.1'

**Comment by hadia206** [comment]
Thanks @auderson for reporting this issue and providing a reproducer. 
I was able to reproduce it as well.
I believe the problem is in the `if-else` statement in `test`  and how the type for `b` is resolved at compile time.
If you replaced your if-else structure to be like the following, it will work
```python
@njit
def test(a):
    if a > 0:
        b = np.arange(5.)
        return inner(np.arange(5.), b)
    else:
        return inner(np.arange(5.))
```

**Comment by auderson** [comment]
@hadia206 

Yes, this will work, as in this version, `b` is not a `OptionalType(array(float64, 1d, C))`. However if I use `if-else` in all the cases to make `b`'s type not optional, my code will be very complicated and redundant.

But luckily, later yesterday I found a way to get around:
just wrap the `y` in `inner` with `np.asarray`:

@njit
def inner(x, x_=None):
    if x_ is not None:
        y = x_
    else:
        y = x + 1
    return np.add(np.asarray(y), 1)

And what interests me here is that can we enable the compiler to infer `y`'s type more precisely, given I've known `x_` is already not `None`, so that it's not of `OptionalType(array(float64, 1d, C))` but `array` type?

**Comment by stuartarchibald** [comment]
This was fixed in #6983
